meta-client/src/main/java/org/apache/helix/metaclient/factories/MetaClientConfig.java

```diff
@@ -0,0 +1,91 @@
+package org.apache.helix.metaclient.factories;
+
+class MetaClientConfig {
+
+  public enum RetryProtocol {
+    NO_RETRY, EXP_BACK_OFF, CONST_RETRY_INTERVAL
+  }
+
+  public enum StoreType {
+    ZOOKEEPER, ETCD
+  }
+
+  private String _connectionAddress;
+  private long _connectionTimeout;
+  private boolean _enableAuth;
+  private StoreType _storeType;
+
+  public String getConnectionAddress() {
+    return _connectionAddress;
+  }
+
+  public long getConnectionTimeout() {
+    return _connectionTimeout;
+  }
+
+  public boolean isAuthEnabled() {
+    return _enableAuth;
+  }
+  public StoreType getStoreType() {
+    return _storeType;
+  }
+
+  // TODO: More options to add later
+  // private boolean _autoReRegistWatcher;  // re-register one time watcher when set to true
+  // private boolean _resetWatchWhenReConnect; // re-register previous existing watcher when reconnect
+  // private RetryProtocol _retryProtocol;
+
+  private MetaClientConfig(Builder builder) {
+    _connectionAddress = builder._connectionAddress;
+    _connectionTimeout = builder._connectionTimeout;
+    _enableAuth = builder._enableAuth;
+    _storeType = builder._storeType;
+
+  }
```

Comment:
Let's input every single variables, but not depends on builder

Reply:
TFTR. Wouldn't input every single variables be the opposite of builder pattern?

Reply:
For builder pattern, my own understanding is that it separates the construction of a complex object from its representation so that the same builder can create different representations with little twist. 
The constructor is still private, and people are only supposed to call builder.build(). The difference is the constructor will not depend on the builder class. Just my two cents.

Reply:
I was reading online for builder pattern and what Qunicy says seems to be the norm.

Reply:
Put it another way, the builder class needs to know how to build the config class, that dependency is required, but the config class doesn't need to depend on the builder. In the new way, the dependency is one-way, not circular.

Reply:
Updated.